Readme.md

# koa-ratelimit

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![node version][node-image]][node-url]

[npm-image]: https://img.shields.io/npm/v/koa-ratelimit.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-ratelimit
[travis-image]: https://img.shields.io/travis/koajs/ratelimit.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/ratelimit
[node-image]: https://img.shields.io/badge/node.js-%3E=_0.11-red.svg?style=flat-square
[node-url]: http://nodejs.org/download/

 Rate limiter middleware for koa 2

## Installation

```js
$ npm install koa-ratelimit@koa2
```

## Example

```js
var ratelimit = require('koa-ratelimit');
var redis = require('redis');
var koa = require('koa');
var app = new Koa();

// apply rate limit

app.use(ratelimit({
  db: redis.createClient(),
  duration: 60000,
  max: 100,
  id: function (context) {
    return context.ip;
  },
  headers: {
    remaining: 'Rate-Limit-Remaining',
    reset: 'Rate-Limit-Reset',
    total: 'Rate-Limit-Total'
  },
  errorMsg: 'Rate limit exceeded, retry in ',
  appendRetryTime: true
}));

// response middleware

app.use(function (ctx, next){
  ctx.body = 'Stuff!';
});

app.listen(3000);
console.log('listening on port 3000');
```

## Options

 - `db` redis connection instance
 - `max` max requests within `duration` [2500]
 - `duration` of limit in milliseconds [3600000]
 - `id` id to compare requests [`(ctx) => (return ctx.ip)`]
 - `headers` custom header names
  - `remaining` remaining number of requests [`'X-RateLimit-Remaining'`]
  - `reset` reset timestamp [`'X-RateLimit-Reset'`]
  - `total` total number of requests [`'X-RateLimit-Limit'`]
  - `retry` retry after time [`'X-Retry-After'`]
 - `throw` throw error if rate limit exceeded [false]
 - `errorMsg` text in body of error response [`'Rate limit exceeded, retry in '`]
 - `appendRetryTime` append retry time to error body text [true]
 - `log` logging function to receive error message [`(ctx, msg) => ()`]

## Responses

  Example 200 with header fields:

```
HTTP/1.1 200 OK
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 99
X-RateLimit-Reset: 1384377793
Content-Type: text/plain; charset=utf-8
Content-Length: 6
Date: Wed, 13 Nov 2013 21:22:13 GMT
Connection: keep-alive

Stuff!
```

  Example 429 response:

```
HTTP/1.1 429 Too Many Requests
X-Powered-By: koa
X-RateLimit-Limit: 100
X-RateLimit-Remaining: 0
X-RateLimit-Reset: 1384377716
Content-Type: text/plain; charset=utf-8
Content-Length: 39
Retry-After: 7
Date: Wed, 13 Nov 2013 21:21:48 GMT
Connection: keep-alive

Rate limit exceeded, retry in 8 seconds
```

## License

  MIT


History.md
3.0.0 / 2016-08-14
==================
  * Release for Koa 2 and other fixes

2.3.0 / 2016-05-12
==================

  * Add support for custom header names (#20)

2.2.0 / 2016-04-11
==================

  * Add option to throw error instead of setting body
  * bump deps

2.1.0 / 2015-01-04
==================

  * Merge pull request #11 from seegno/support/flush-redis-before-testing
  * Merge pull request #9 from seegno/enhancement/add-id-verification
  * Flush redis limit keys before running tests
  * Add id verification to bypass ratelimiting

2.0.0 / 2015-01-02
==================

  * Merge pull request #10 from seegno/support/update-ratelimiter-2-0-0
  * Update `debug@2.1.1` and `ms@0.7.0`
  * Update recommended node version to >= 0.11.13
  * Update dev dependencies
  * Update ratelimiter@2.0.0
  * Replace `thunkify` by `thenify`.

1.1.0 / 2014-12-03
==================

  * Adds callback for defining own comparator ID. Fixes #3

1.0.3 / 2014-05-24
==================

 * fix: don't yield to downstream middleware if ratelimit is hit.

1.0.2 / 2014-05-22
==================

 * yield to next only when rate limit isn't hit

1.0.1 / 2013-11-29
==================

 * fix ms dep


test/ratelimit.js

var request = require('supertest');
var should = require('should');
var redis = require('redis');
var Koa = require('koa');

var ratelimit = require('..');

var db = redis.createClient();

describe('ratelimit middleware', function() {
  var rateLimitDuration = 1000;
  var goodBody = "Num times hit: ";

  before(function(done) {
    db.keys('limit:*', function(err, rows) {
      rows.forEach( (element) => {
        db.del(element, () => {});  
      });
    });

    done();
  });

  describe('limit', function() {
    var guard;
    var app;

    var routeHitOnlyOnce = function() {
      guard.should.be.equal(1);
    };

    beforeEach(function(done) {
      app = new Koa();

      app.use(ratelimit({
        duration: rateLimitDuration,
        db: db,
        max: 1
      }));

      app.use(function (ctx, next) {
        guard++;
        ctx.body = goodBody + guard;
      });

      guard = 0;

      setTimeout(function() {
        request(app.listen())
          .get('/')
          .expect(200, goodBody + "1")
          .expect(routeHitOnlyOnce)
          .end(done);
      }, rateLimitDuration);
    });

    it('responds with 429 when rate limit is exceeded', function(done) {
      request(app.listen())
        .get('/')
        .expect('X-RateLimit-Remaining', 0)
        .expect(429)
        .end(done);
    });

    it('should not yield downstream if ratelimit is exceeded', function(done) {
      request(app.listen())
        .get('/')
        .expect(429)
        .end(function() {
          routeHitOnlyOnce();
          done();
        });
    });
  });

  describe('limit with throw', function() {
    var guard;
    var app;

    var routeHitOnlyOnce = function() {
      guard.should.be.equal(1);
    };

    beforeEach(function(done) {
      app = new Koa();

      app.use(function (ctx, next) {
        try {
          return next();
        } catch (e) {
          ctx.body = e.message;
          ctx.set(e.headers);
        }
      });

      app.use(ratelimit({
        duration: rateLimitDuration,
        db: db,
        max: 1,
        throw: true
      }));

      app.use(function (ctx, next) {
        guard++;
        ctx.body = goodBody + guard;
      });

      guard = 0;

      setTimeout(function() {
        request(app.listen())
          .get('/')
          .expect(200, goodBody + "1")
          .expect(routeHitOnlyOnce)
          .end(done);
      }, rateLimitDuration);
    });

    it('responds with 429 when rate limit is exceeded', function(done) {
      request(app.listen())
        .get('/')
        .expect('x-rateLimit-remaining', 0)
        .expect(429)
        .end(done);
    });
  });

  describe('id', function (done) {
    it('should allow specifying a custom `id` function', function (done) {
      var app = new Koa();

      app.use(ratelimit({
        db: db,
        max: 1,
        id: function (ctx) {
          return ctx.request.header.foo;
        }
      }));

      request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect(function(res) {
          res.header['x-ratelimit-remaining'].should.equal('0');
        })
        .end(done);
    });

    it('should not limit if `id` returns `false`', function (done) {
      var app = new Koa();

      app.use(ratelimit({
        db: db,
        id: function (ctx) {
          return false;
        },
        max: 5
      }));

      request(app.listen())
        .get('/')
        .expect(function(res) {
          res.header.should.not.have.property('x-ratelimit-remaining');
        })
        .end(done);
    });

    it('should limit using the `id` value', function (done) {
      var app = new Koa();

      app.use(ratelimit({
        db: db,
        max: 1,
        id: function (ctx) {
          return ctx.request.header.foo;
        }
      }));

      app.use(function (ctx, next) {
        ctx.body = ctx.request.header.foo;
      });

      request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect(200, 'bar')
        .end(function() {
          request(app.listen())
            .get('/')
            .set('foo', 'biz')
            .expect(200, 'biz')
            .end(done);
        });
    });
  });

  
  describe('errorMsg', function (done) {
    it('should allow using a custom error body message using the `errorMsg` value', function (done) {
      var app = new Koa();

      app.use(ratelimit({
        db: db,
        max: 1,
        errorMsg: 'Exceeded limit, retry in '
      }));

      request(app.listen())
        .get('/')
        .expect(429)
        .expect(function(res) {
          res.text.should.startWith('Exceeded limit, retry in');
          res.text.should.not.startWith('Rate limit exceeded, retry in');
        })
        .end(done);
    });
  });
  
  describe('custom headers', function() {
    it('should allow specifying a custom header names', function(done) {
      var app = new Koa();

      app.use(ratelimit({
        db: db,
        max: 1,
        headers: {
          remaining: 'Rate-Limit-Remaining',
          reset: 'Rate-Limit-Reset',
          total: 'Rate-Limit-Total'
        }
      }));

      request(app.listen())
        .get('/')
        .set('foo', 'bar')
        .expect(function(res) {
          res.headers.should.containEql('rate-limit-remaining', 'rate-limit-reset', 'rate-limit-total');
          res.headers.should.not.containEql('x-ratelimit-limit', 'x-ratelimit-remaining', 'x-ratelimit-reset');
        })
        .end(done);
    });
  });
  
  describe('logging function', function() {
    var app, logMsg, logStatus;

    beforeEach(function(done) {
      app = new Koa();
      logMsg = null;
      logStatus = 500;

      app.use(ratelimit({
        db: db,
        errorMsg: 'ERROR_RATE_LIMIT_EXCEEDED',
        appendRetryTime: false,
        throw: false,
        log: function(ctx, msg) { logMsg = msg; logStatus = ctx.status; }
      }));
  
      
      app.use(function (ctx, next) {
        ctx.body = 'Not blocked!';
      });
      
      done();
    });
    
    it('should log out correct error message', function(done) {
      request(app.listen())
        .get('/')
        .expect(429)
        .expect( (res) => {
          logMsg.should.equal('ERROR_RATE_LIMIT_EXCEEDED');
          logStatus.should.equal(429);
        })
        .end(done);
    });
    
  });
});


index.js

/**
 * Module dependencies.
 */

var debug = require('debug')('koa-ratelimit');
var Limiter = require('ratelimiter');
var ms = require('ms');
var thenify = require('thenify');

/**
 * Expose `ratelimit()`.
 */

module.exports = ratelimit;

/**
 * Initialize ratelimit middleware with the given `opts`:
 *
 * - `duration` limit duration in milliseconds [1 hour]
 * - `max` max requests per `id` [2500]
 * - `db` database connection
 * - `id` id to compare requests [ (ctx) => (return ctx.ip) ]
 * - `headers` custom header names
 *  - `remaining` remaining number of requests ['X-RateLimit-Remaining']
 *  - `reset` reset timestamp ['X-RateLimit-Reset']
 *  - `total` total number of requests ['X-RateLimit-Limit']
 *  - `retry` retry after time ['X-Retry-After']
 * - `errorMsg` text in body of error response ['Rate limit exceeded, retry in ']
 * - `appendRetryTime` append retry time to error body text [true]
 * - `log` logging function to receive error message [(ctx, msg) => ()]
 *
 * @param {Object} opts
 * @return {Function}
 * @api public
 */

function ratelimit(opts) {
  opts = opts || {};
  opts.headers = opts.headers || {};
  opts.headers.remaining = opts.headers.remaining || 'X-RateLimit-Remaining';
  opts.headers.reset = opts.headers.reset || 'X-RateLimit-Reset';
  opts.headers.total = opts.headers.total || 'X-RateLimit-Limit';
  opts.headers.retry = opts.headers.retry || 'X-Retry-After';
  opts.errorMsg = opts.errorMsg || 'Rate limit exceeded, retry in ';
  opts.appendRetryTime = (typeof opts.appendRetryTime !== 'undefined') ? opts.appendRetryTime : true;
  opts.throw = (typeof opts.throw !== 'undefined') ? opts.throw : false;

  return function (ctx, next){
    var id = opts.id ? opts.id(ctx) : ctx.ip;

    if (false === id) return next();

    // initialize limiter
    var limiter = new Limiter({ id: id, __proto__: opts });
    limiter.get = thenify(limiter.get);

    // check limit
    return limiter.get().then((limit) => {
      // check if current call is legit
      var remaining = limit.remaining > 0 ? limit.remaining - 1 : 0;

      // header fields
      var headers = {};
      headers[opts.headers.remaining] = remaining;
      headers[opts.headers.reset] = limit.reset;
      headers[opts.headers.total] = limit.total;

      ctx.set(headers);

      debug('remaining %s/%s %s', remaining, limit.total, id);
      if (limit.remaining) return next();

      var delta = (limit.reset * 1000) - Date.now() | 0;
      var after = limit.reset - (Date.now() / 1000) | 0;
      ctx.set(opts.headers.retry, after);

      ctx.status = 429;
      ctx.body = opts.errorMsg + (opts.appendRetryTime ? ms(delta, { long: true }) : '');
      
      if (opts.log) opts.log(ctx, opts.errorMsg);

      if (opts.throw) ctx.throw(ctx.status, ctx.body, { headers: headers });
    });
  }
}
